internal/pkg/service/common/etcdop/op/atomic.go

```diff
@@ -56,8 +56,13 @@ type Mutex interface {
 func Atomic[R any](client etcd.KV, result *R) *AtomicOp[R] {
 	v := &AtomicOp[R]{AtomicOpCore: newAtomicCore(client), result: result}
 	v.setProcessorFactory(func() func(ctx context.Context, r *TxnResult[NoResult]) {
+		// Invoke processors only if the transaction succeeded and there is no collision.
+		// Otherwise, succeeded == false, the processors are skipped,
+		// because the atomic operation is retried due to a collision or an etcd error, see AtomicOp.Do method.
 		return func(ctx context.Context, r *TxnResult[NoResult]) {
-			v.processors.invoke(ctx, newResult(r.Response(), v.result))
+			if r.Err() != nil || r.Succeeded() {
+				v.processors.invoke(ctx, newResult(r.Response(), v.result))
+			}
```

Comment:
Reverted/fixed :man_facepalming: 

I didn't remember why it was there and I deleted the condition and didn't run all the tests.

Fixed and I added the comment.

See comment: 
https://github.com/keboola/keboola-as-code/pull/1767#discussion_r1604499784

AtomicOp is retried here:
https://github.com/keboola/keboola-as-code/blob/c7f5e796aa1178d25c7c3afb25e72ceefc612329/internal/pkg/service/common/etcdop/op/atomic_do.go#L25

------------

It is questionable whether there should/shouldn't be `r.Err != nil`. it means, does it make sense for the processor to solve some network and similar errors.

Reply:
cc @jachym-tousek-keboola @hosekpeter 

-----

Etcd client itself has retry mechanism, but only for operations whose retried execution will certainly not cause a problem, e.g. `GET RANGE`. Retries are not made for `PUT`, because it can lead to the fact that the key will have 2 revisions, two historical versions instead of one. If the first attempt was successful, but the report did not reach the client.

-----

For all operations, we have additional retry mechanism for etcd/network errors, see:
https://github.com/keboola/keboola-as-code/blob/02d3bde2cd901e0c375008edbc41f8bdfa3395ee/internal/pkg/service/common/etcdop/op/op.go#L84-L88

https://github.com/keboola/keboola-as-code/blob/c1adb0608219fe3fb7610e145b9bd733b1a9ffb8/internal/pkg/service/common/etcdop/op/retry.go#L35-L72

We assume following errors as safe for retry, for our use-cases.
https://github.com/keboola/keboola-as-code/blob/c1adb0608219fe3fb7610e145b9bd733b1a9ffb8/internal/pkg/service/common/etcdop/op/retry.go#L16-L28

-------------------

AND! we have additional retry mechansim for `AtomicOp`, primarily for a logical :exclamation:  error when a collision occurred and the top-level IF conditions of `AtomicOp` were not met.

See `!ok`:
https://github.com/keboola/keboola-as-code/blob/8fd94c901089f8c17d58385111a7f0924c78bde0/internal/pkg/service/common/etcdop/op/atomic_do.go#L21-L35